src/data_structures/interval_tree.rs

```diff
@@ -0,0 +1,359 @@
+extern crate num;
+
+use self::num::traits::Num;
+
+use std::cmp;
+use std::mem;
+use std::fmt::Debug;
+use std::ops::Range;
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub struct IntervalTree<N, D> {
+    root: Option<Node<N, D>>,
+}
+
+impl<N: Debug + Num + Clone + Ord, D: Debug> IntervalTree<N, D> {
+    pub fn new() -> Self {
+        IntervalTree { root: None }
+    }
+
+    pub fn insert(&mut self, interval: Range<N>, data: D) {
+        validate(&interval);
+        match self.root {
+            Some(ref mut n) => n.insert(interval, data),
+            None => self.root = Some(Node::new(interval, data)),
+        }
+    }
+
+    pub fn find_any(&self, interval: &Range<N>) -> Option<(&Range<N>, &D)> {
+        validate(&interval);
+        match self.root {
+            Some(ref n) => {
+                let mut result = vec![];
+                n.find(interval, true, &mut result);
+                result.pop()
+            }
+            None => None,
+        }
+    }
+
+    pub fn find_all(&self, interval: &Range<N>) -> Vec<(&Range<N>, &D)> {
```

Comment:
Maybe yield  structs `Entry` here. This is easier to use, e.g. via `entry.interval` and `entry.data`.